ceph-volume: allow encryption per device type

clwluvw opened the pull request:
Extend `--dmcrypt` flag to support per device type encryption if specified.

Fixes: https://tracker.ceph.com/issues/64798

clwluvw committed:
ceph-volume: allow encryption per device type

Extend --dmcrypt flag to support per device type encryption if specified.

Fixes: https://tracker.ceph.com/issues/64798
Signed-off-by: Seena Fallah <seenafallah@gmail.com>

clwluvw commented:
jenkins test this please

guits commented on src/ceph-volume/ceph_volume/devices/raw/prepare.py:
```diff
@@ -90,10 +90,10 @@ def safe_prepare(self, args=None):
         secrets = {'cephx_secret': prepare_utils.create_key()}
-        encrypted = 1 if self.args.dmcrypt else 0
-        cephx_lockbox_secret = '' if not encrypted else prepare_utils.create_key()
+        encrypt_devices = self.args.dmcrypt
```
I'd suggest a better terminology:

```suggestion
        encrypted_device_types = self.args.dmcrypt
```

guits commented on src/ceph-volume/ceph_volume/devices/lvm/prepare.py:
```diff
@@ -256,30 +255,35 @@ def prepare(self):
             tags['ceph.vdo'] = api.is_vdo(block_lv.lv_path)
 
+            def set_encryption_tag(tags, device_type):
+                tags.update({'ceph.encrypted': 1 if device_type in encrypt_devices else 0})
```
```python
if device_type in encrypt_devices
```
this will fail when `--dmcrypt` is not passed as `self.args.dmcrypt` will be `None`

clwluvw committed:
ceph-volume: use _AppendAction for DmcryptAction argparse

argparse._ExtendAction was introduced in python 3.8. Currently upstream still maintains python 3.6 which doesn't have argparse._ExtendAction supported. This commit can be reverted after upgraded to python 3.8+.

Signed-off-by: Seena Fallah <seenafallah@gmail.com>

guits commented on src/ceph-volume/ceph_volume/util/arg_validators.py:
```diff
@@ -10,13 +11,27 @@
+        if not values:
+            values = ['block', 'db', 'wal']
+
+        # TODO: use argparse._ExtendAction when upgraded to python 3.8
```
```suggestion
        # TODO: use argparse._ExtendAction when upgraded to python >= 3.8
```

guits commented:
jenkins test api

guits commented:
jenkins test make check

clwluvw commented:
unstale please

clwluvw commented:
unstale please

clwluvw commented:
unstale please
